### PR TITLE
Adds assert_byte_buf_equivalence function to proof helpers

### PR DESCRIPTION
**Pull request by feliperodri**
Signed-off-by: Felipe R. Monteiro <felisous@amazon.com>

*Description of changes:*
Adds `assert_byte_buf_equivalence` function.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by feliperodri** [comment]
@karkhaz @danielsn any comments?

File: .cbmc-batch/source/utils.c [hunk 1]
```diff
@@ -64,6 +64,18 @@ void assert_array_list_equivalence(
     }
 }
 
+void assert_byte_buf_equivalence(
+    const struct aws_byte_buf *const lhs,
+    const struct aws_byte_buf *const rhs,
+    const struct store_byte_from_buffer *const rhs_byte) {
+    assert(lhs->len == rhs->len);
```
Comment (karkhaz) [inline .cbmc-batch/source/utils.c]: Don't you need to check that `lhs` and `rhs` are not null before you dereference `len`?

Comment (feliperodri) [reply .cbmc-batch/source/utils.c]: Yes, thank you!

File: .cbmc-batch/include/proof_helpers/utils.h [hunk 2]
```diff
@@ -64,6 +65,16 @@ void assert_array_list_equivalence(
     const struct aws_array_list *const rhs,
     const struct store_byte_from_buffer *const rhs_byte);
 
+/**
```
Comment (danielsn) [inline .cbmc-batch/include/proof_helpers/utils.h]: See comment on previous review.

File: .cbmc-batch/source/utils.c [hunk 3]
```diff
@@ -38,7 +38,7 @@ void assert_all_zeroes(const uint8_t *const a, const size_t len) {
 }
 
 void assert_byte_from_buffer_matches(const uint8_t *const buffer, const struct store_byte_from_buffer *const b) {
-    if (buffer) {
+    if (buffer && b) {
```
Comment (danielsn) [inline .cbmc-batch/source/utils.c]: when would we ever not have a b?

File: .cbmc-batch/source/utils.c [hunk 4]
```diff
@@ -55,12 +55,30 @@ void assert_array_list_equivalence(
     const struct aws_array_list *const lhs,
     const struct aws_array_list *const rhs,
     const struct store_byte_from_buffer *const rhs_byte) {
-    assert(lhs->alloc == rhs->alloc);
-    assert(lhs->current_size == rhs->current_size);
-    assert(lhs->length == rhs->length);
-    assert(lhs->item_size == rhs->item_size);
-    if (lhs->current_size > 0) {
-        assert_byte_from_buffer_matches((uint8_t *)lhs->data, rhs_byte);
+    assert(!lhs == !rhs);
```
Comment (danielsn) [inline .cbmc-batch/source/utils.c]: Clever.  But a comment might help

File: .cbmc-batch/source/utils.c [hunk 5]
```diff
@@ -55,12 +55,30 @@ void assert_array_list_equivalence(
...
+            assert_byte_from_buffer_matches((uint8_t *)lhs->data, rhs_byte);
+        }
+    }
+}
+
+void assert_byte_buf_equivalence(
+    const struct aws_byte_buf *const lhs,
+    const struct aws_byte_buf *const rhs,
+    const struct store_byte_from_buffer *const rhs_byte) {
+    assert(!lhs == !rhs);
```
Comment (danielsn) [inline .cbmc-batch/source/utils.c]: Comment here.

File: .cbmc-batch/source/utils.c [hunk 6]
```diff
@@ -55,12 +55,30 @@ void assert_array_list_equivalence(
...
+    if (lhs && rhs) {
```
Comment (danielsn) [inline .cbmc-batch/source/utils.c]: If we have ```lhs``` but not ```rhs```, what are the sematics.

File: .cbmc-batch/source/utils.c (hunk 3 above)
Comment (feliperodri) [reply .cbmc-batch/source/utils.c]: When one tries to save a byte from a NULL buffer using `save_byte_from_array`

File: .cbmc-batch/source/utils.c (hunk 6 above)
Comment (feliperodri) [reply .cbmc-batch/source/utils.c]: What do you mean? The assertion on the previous line ensures either both are null or both are non-null. Could you clarify the comment?

Comment (danielsn) [reply .cbmc-batch/source/utils.c]:
```
if (lhs == rhs) return;
assert (lhs && rhs) // if only one is null, they're different
//they're both non-null, now, check the internals.
```

**Pull request merged**